target_chains/solana/programs/pyth-price-publisher/src/accounts/publisher_prices.rs

```diff
@@ -0,0 +1,173 @@
+};
+
+/// Account Magic to avoid Account Confusiong
+const FORMAT: u32 = 2848712303;
+
+#[derive(Debug, Clone, Copy, Zeroable, Pod)]
+#[repr(C, packed)]
+pub struct PublisherPricesHeader {
+    pub format: u32,
+    pub publisher: [u8; 32],
+    pub slot: Slot,
+    pub num_prices: u32,
+}
+
+impl PublisherPricesHeader {
+    pub fn new(publisher: [u8; 32]) -> Self {
+        PublisherPricesHeader {
+            format: FORMAT,
+            publisher,
+            slot: 0,
+            num_prices: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Zeroable, Pod)]
+#[repr(C, packed)]
+pub struct PublisherPrice {
+    // 4 high bits: trading status
+    // 28 low bits: feed index
+    pub trading_status_and_feed_index: u32,
+    pub price: i64,
+    pub confidence: u64,
+}
+
+#[derive(Debug, Error)]
+#[error("publisher price data overflow")]
+pub struct PublisherPriceError;
+
+impl PublisherPrice {
+    pub fn new(
+        feed_index: u32,
+        trading_status: u32,
+        price: i64,
+        confidence: u64,
+    ) -> Result<Self, PublisherPriceError> {
+        if feed_index >= (1 << 28) || trading_status >= (1 << 4) {
+            return Err(PublisherPriceError);
+        }
+        Ok(Self {
+            trading_status_and_feed_index: (trading_status << 28) | feed_index,
+            price,
+            confidence,
+        })
+    }
+
+    pub fn trading_status(&self) -> u32 {
+        self.trading_status_and_feed_index >> 28
+    }
+
+    pub fn feed_index(&self) -> u32 {
+        self.trading_status_and_feed_index & ((1 << 28) - 1)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ReadAccountError {
+    #[error("data too short")]
+    DataTooShort,
+    #[error("format mismatch")]
+    FormatMismatch,
+    #[error("invalid num prices")]
+    InvalidNumPrices,
+}
+
+#[derive(Debug, Error)]
+pub enum ExtendError {
+    #[error("not enough space")]
+    NotEnoughSpace,
+    #[error("invalid length")]
+    InvalidLength,
+}
+
+pub fn format_matches(data: &[u8]) -> bool {
+    if data.len() < size_of::<u32>() {
+        return false;
+    }
+    let format: &u32 = from_bytes(&data[..size_of::<u32>()]);
+    *format == FORMAT
+}
+
+pub fn read(data: &[u8]) -> Result<(&PublisherPricesHeader, &[PublisherPrice]), ReadAccountError> {
+    if data.len() < size_of::<PublisherPricesHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let header: &PublisherPricesHeader = from_bytes(&data[..size_of::<PublisherPricesHeader>()]);
+    if header.format != FORMAT {
+        return Err(ReadAccountError::FormatMismatch);
+    }
+    let prices_bytes = &data[size_of::<PublisherPricesHeader>()..];
+    let num_prices: usize = header.num_prices.try_into().unwrap();
+    let expected_len = num_prices.saturating_mul(size_of::<PublisherPrice>());
+    if expected_len > prices_bytes.len() {
+        return Err(ReadAccountError::InvalidNumPrices);
+    }
+    let prices = cast_slice(&prices_bytes[..expected_len]);
```

Comment:
please add a note here too that we don't do validation here.

Reply:
Done.